src/django_otp/models.py

```diff
@@ -226,10 +225,28 @@ class SideChannelDevice(Device):
         default=timezone.now,
         help_text="The timestamp of the moment of expiry of the saved token.",
     )
+    last_generated = models.DateTimeField(
```

Comment:
This will trigger a migration in existing devices, which may require a semantic version bump. Is it possible for it to live in `GenerationThrottlingMixin`? Also, should it get reset on a successful validation?

Reply:
> This will trigger a migration in existing devices, which may require a semantic version bump

I believe it belongs together with the rest of the token attributes (`token` and `valid_until`). Indeed it will trigger a migration, but since it is `null=True`, I don't think it will break anything on existing installations. Therefore, it would not need a major version bump, unless of course that is your policy.

> Is it possible for it to live in GenerationThrottlingMixin?

In my opinion (for the reasons listed in my first comment), this mixin should be the default for the interactive devices. That includes the EmailDevice that ships with the package. Therefore, that would trigger a migration anyway.

> Also, should it get reset on a successful validation?
Indeed, I will add this.

On a final note, I was thinking how we can do this without any migration and one approach is to use the `valid_until`. Which means, the user can generate another token, only if the previous token has expired. However, I decided that it is better to have a different expiration for generation. But I mentioning it here, in case you prefer to go this way.

Reply:
Suppose django-otp adds a field to SideChannelDevice, but some other project similar to django-otp-twilio doesn't get updated for a little while? What happens when an unsuspecting project gets a new field on a third-party model that they have no direct control over?

Reply:
> Suppose django-otp adds a field to SideChannelDevice, but some other project similar to django-otp-twilio doesn't get updated for a little while? What happens when an unsuspecting project gets a new unmigrated field on a third-party model that they have no direct control over?

Sorry if this is trivial, but I am trying to follow this through; the user installs django-otp which includes the new field in SideChannelDevice. Also installs a project which is unaware of this change e.g. "django-otp-twillio". Since it uses `SideChannelDevice` model from django-otp, this will trigger a migration for the `TwilioSMSDevice`. The migration can be applied without having any effect on the application. What am I missing?

Reply:
Let's think through what actually happens here. You have a project with a dependency on django-otp-twilio (in this example). You update django-otp and run makemigrations. Django sees that TwilioSMSDevice has acquired a new field and needs a migration. Where is it going to write this migration file? In the installed django-otp-twilio package in your development virtualenv? How does that help us?

Even if you could somehow generate a migration for a third-party app and check it in somewhere, what happens when django-otp-twilio updates itself and adds the official migration with its own name? You update again and migrate and Django tries to apply a new migration to add a field that already exists in your database table. Now I guess you have to `migrate --fake` or something. Total chaos.

I haven't tried simulating this, but I assume that's what would happen. The burden of proof is on demonstrating conclusively that this can be made to work seamlessly and transparently.

Reply:
> Even if you could somehow generate a migration for a third-party app and check it in somewhere, what happens when django-otp-twilio updates itself and adds the official migration with its own name? You update again and migrate and Django tries to apply a new migration to add a field that already exists in your database table. Now I guess you have to `migrate --fake` or something. Total chaos.

Indeed that is an undesirable case and should be avoided. 

1. I am curious, under the same terms, does that mean that none of the existing models can be altered, ever? 
2. Does this mean that the whole PR falls apart? Because even if the migration is triggered by the mixin, the same case applies, if the EmailDevice uses the mixin by default.

Reply:
> Is it possible for it to live in GenerationThrottlingMixin?

Completed

> Also, should it get reset on a successful validation?

Completed.